File: phSolver/common/test/phIOwrite.cc
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include "phIO.h"
#include "syncio.h"
#include "posixio.h"

int main(int argc, char* argv[]) {
  MPI_Init(&argc,&argv);
  int rank;
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  int size;
  MPI_Comm_size(MPI_COMM_WORLD, &size);
  if( argc != 2 ) {
    fprintf(stderr, "Usage: %s <numSyncFiles>\n",argv[0]);
    MPI_Finalize();
    return 1;
  }
  const char* phrase = "number of fishes";
  const char* type = "double";
  const char* iotype = "binary";
  int one = 1;
  int fish = 2;
  int numFish[2] = {0,0};
  double fishWeight[2] = {1.23,1.23};
  int nfiles = atoi(argv[1]);
  int ppf = size/nfiles;
  const char* filename[2] = {"water-dat.", "water.dat."};
  phio_fp file[2];
  syncio_setup_write(nfiles, one, ppf, &(file[0]));
  posixio_setup(&(file[1]), 'w');
  for(int i=0; i<2; i++) {
    phio_openfile(filename[i], file[i]);
    phio_writeheader(file[i], phrase, &fish, &one, &one, type, iotype);
    phio_writedatablock(file[i], phrase, &(fishWeight[i]), &one, type, iotype);
    phio_closefile(file[i]);
  }
  syncio_setup_read(nfiles, &(file[0]));
  posixio_setup(&(file[1]), 'r');
  for(int i=0; i<2; i++) {
    phio_openfile(filename[i], file[i]);
    phio_readheader(file[i], phrase, &(numFish[i]), &one, type, iotype);
    phio_readdatablock(file[i], phrase, &(fishWeight[i]), &one, type, iotype);
    phio_closefile(file[i]);
  }
  int match = (numFish[0] == numFish[1]) && (fishWeight[0] == fishWeight[1]);
  if(!rank && match)
    fprintf(stderr, "number of fish match!\n");
  if(!rank && !match)
    fprintf(stderr, "number of fish don't match... :(\n");
  MPI_Finalize();
  return !match;
}

File: phSolver/common/test/phIOwrite.f
      program readheaderFtn
      use iso_c_binding
      use phio
      use syncio
      use posixio
      include "mpif.h"

      integer, target :: rank, ierror, one, ppf, peers, fish, nfiles
      type(c_ptr), dimension(2) :: handle
      character(len=30) :: dataDbl, iotype
      character(len=256) :: phrase
      character(len=256), dimension(2) :: fname
      integer, target, dimension(2) :: fishweight, numFish

      call MPI_Init(ierror)
      call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierror)
      call MPI_Comm_size(MPI_COMM_WORLD, peers, ierror)

      phrase = c_char_"number of fishes"//c_null_char
      dataDbl = c_char_"double"//c_null_char
      iotype =  c_char_"binary"//c_null_char
      one = 1 
      fish = 2

      fishweight(1) = 1.23
      fishweight(2) = 1.23

      fname(1) = c_char_"fortranWater-dat."//c_null_char
      fname(2) = c_char_"fortranWater.dat."//c_null_char
      nfiles = 2
      ppf = peers/nfiles
      ! handle(1) is the file for syncio writing
      call syncio_setup_write(nfiles, one, ppf, handle(1))
      ! handle(2) is the file for posix writing
      call posixio_setup(handle(2), c_char_"w"//c_null_char)
      ! if there were a handle(3) for streams we would do the following
      ! call streamio_setup_write(handle(3), <stream obj>)
      ! after the handles are setup the function calls are the same
      ! write the same garbage to posix and syncio files
      do i=1,2
        call phio_openfile(fname(i), handle(i))
        call phio_writeheader(handle(i), phrase, c_loc(fish), one, one,
     &      dataDbl, iotype)
        call phio_writedatablock(handle(i), phrase, c_loc(fishweight(i)),
     &      one, dataDbl, iotype)
        call phio_closefile(handle(i))
      end do
      ! this is the read side... less interesting for us
      call syncio_setup_read(nfiles, handle(1))
      call posixio_setup(handle(2), c_char_"r"//c_null_char)
      do i=1,2
        call phio_openfile(fname(i), handle(i))
        call phio_readheader(handle(i), phrase, c_loc(numFish(i)),
     &      one, dataDbl, iotype)
        call phio_readdatablock(handle(i), phrase, c_loc(fishweight(i)),
     &      one, dataDbl, iotype)
        call phio_closefile(handle(i))
      end do
      if( numFish(1) .ne. numFish(2) .or.
     &    fishweight(1) .ne. fishweight(2) ) then
        write (*,*) "fish don\'t match"
        stop 1
      end if
      call MPI_Finalize(ierror)
      end
